README.md
# mm-code README

This is the README for your extension "mm-code". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

-----------------------------------------------------------------------------------------------------------

## Working with Markdown

**Note:** You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on OSX or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+CMD+V` on OSX or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux) or `Cmd+Space` (OSX) to see a list of Markdown snippets

### For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**

CHANGELOG.md
# Change Log
All notable changes to the "mm-code" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

out/extension.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const vscode_1 = require("vscode");
const timers_1 = require("timers");
const WebSocket = require('ws');
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    let controller = new WordCounterController();
    context.subscriptions.push(controller);
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    context.subscriptions.push(vscode.commands.registerCommand('extension.mmCodeStart', () => {
        controller.startConnect();
    }));
    context.subscriptions.push(vscode.commands.registerCommand('extension.mmCodeStop', () => {
        controller.stopConnect();
        vscode.window.showInformationMessage('MM stopped.');
    }));
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;
class WordCounterController {
    constructor() {
        this._statusBarItem = vscode_1.window.createStatusBarItem(vscode_1.StatusBarAlignment.Left);
        this._statusBarItem.show();
        this._setStateRun();
    }
    dispose() {
        this.stopConnect();
        this._statusBarItem.dispose();
    }
    stopConnect() {
        if (this._client) {
            this._client.close();
            this._client = null;
        }
        if (this._disposable) {
            this._disposable.dispose();
            this._disposable = null;
        }
        if (this._retryLoop) {
            clearTimeout(this._retryLoop);
            this._retryLoop = null;
        }
        this._setStateRun();
    }
    startConnect() {
        this.stopConnect();
        let subscriptions = [];
        vscode_1.window.onDidChangeTextEditorSelection(this._onDidChangeTextEditorSelection, this, subscriptions);
        vscode_1.window.onDidChangeActiveTextEditor(this._onDidChangeActiveTextEditor, this, subscriptions);
        vscode_1.workspace.onDidChangeTextDocument(this._onDidChangeTextDocument, this, subscriptions);
        this._disposable = vscode_1.Disposable.from(...subscriptions);
        this._client = new WebSocket('ws://localhost:8090/ws');
        this._client.onerror = (e) => {
            console.log('Connection Error', e);
            vscode.window.showErrorMessage('Failed to connect MM server. MM will retry after 3 sec.');
            if (this._retryLoop) {
                clearTimeout(this._retryLoop);
                this._retryLoop = null;
            }
            this._retryLoop = timers_1.setTimeout(() => {
                this.startConnect();
            }, 1000 * 3);
        };
        this._client.onopen = () => {
            console.log('WebSocket Client Connected');
            vscode.window.showInformationMessage('MM Client Connected');
            this._client.send('join ' + JSON.stringify({
                name: 'room'
            }));
            this._putLatestFile();
        };
        this._client.onclose = () => {
            console.log('WebSocket Client Closed');
        };
        this._client.onmessage = (e) => {
            const data = JSON.parse(e.data);
            const type = data.type;
            if (type === 'join') {
                // 誰か入室
                this._putLatestFile();
            }
        };
        this._setStateStop();
    }
    _setStateRun() {
        this._statusBarItem.text = "Run MM";
        this._statusBarItem.command = "extension.mmCodeStart";
    }
    _setStateStop() {
        this._statusBarItem.text = "Stop MM";
        this._statusBarItem.command = "extension.mmCodeStop";
    }
    _getFileData() {
        // アクティブなエディタを取得
        const editor = vscode_1.window.activeTextEditor;
        return {
            fileName: editor ? editor.document.fileName : '',
            text: editor ? editor.document.getText() : ''
        };
    }
    _putLatestFile(id = "") {
        const data = this._getFileData();
        this._client.send('follow ' + JSON.stringify({
            id: id,
            fileName: data.fileName,
            text: data.text,
        }));
        this._putLine();
    }
    _putLine() {
        // アクティブなエディタを取得
        const editor = vscode_1.window.activeTextEditor;
        if (editor) {
            const lineIndex = editor.selection.anchor.line;
            const columnIndex = editor.selection.anchor.character;
            const lineText = editor.document.lineAt(lineIndex).text;
            this._client.send('line ' + JSON.stringify({
                r: lineIndex,
                c: columnIndex,
                l: lineText
            }));
        }
    }
    _onDidChangeTextDocument(e1) {
        const updates = [];
        e1.contentChanges.forEach((e2) => {
            updates.push({
                sr: e2.range.start.line,
                sc: e2.range.start.character,
                er: e2.range.end.line,
                ec: e2.range.end.character,
                t: e2.text
            });
        });
        this._client.send('updates ' + JSON.stringify({
            updates: updates
        }));
    }
    _onDidChangeActiveTextEditor(e) {
        this._putLatestFile();
    }
    _onDidChangeTextEditorSelection(e) {
        this._putLine();
    }
}
//# sourceMappingURL=extension.js.map

src/extension.ts
'use strict'
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode'
import {window, commands, Disposable, ExtensionContext, StatusBarAlignment, StatusBarItem, TextDocument, workspace} from 'vscode'
import { setTimeout } from 'timers'
const WebSocket = require('ws')
import * as child_process from 'child_process'

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
    let controller = new WordCounterController()
    context.subscriptions.push(controller)

    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    context.subscriptions.push(vscode.commands.registerCommand('extension.mmCodeStart', () => {
        controller.startConnect()
    }))

    context.subscriptions.push(vscode.commands.registerCommand('extension.mmCodeStop', () => {
        controller.stopConnect()
        vscode.window.showInformationMessage('MM stopped.')
    }))
}

// this method is called when your extension is deactivated
export function deactivate() {
}

class WordCounterController {
    
        private _disposable: Disposable
        private _statusBarItem: StatusBarItem
        private _client
        private _retryLoop: NodeJS.Timer
    
        constructor() {
            this._statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left)
            this._statusBarItem.show()
            this._setStateRun()
        }
    
        dispose() {
            this.stopConnect()
            this._statusBarItem.dispose()
        }

        stopConnect() {
            if (this._client) {
                this._client.close()
                this._client = null
            }
            if (this._disposable) {
                this._disposable.dispose()
                this._disposable = null
            }
            if (this._retryLoop) {
                clearTimeout(this._retryLoop)
                this._retryLoop = null
            }
            this._setStateRun()
        }

        startConnect() {
            this.stopConnect()

            let subscriptions: Disposable[] = []
            window.onDidChangeTextEditorSelection(this._onDidChangeTextEditorSelection, this, subscriptions)
            window.onDidChangeActiveTextEditor(this._onDidChangeActiveTextEditor, this, subscriptions)
            workspace.onDidChangeTextDocument(this._onDidChangeTextDocument, this, subscriptions)
            this._disposable = Disposable.from(...subscriptions)
            
            this._client = new WebSocket('ws://localhost:8090/ws')
            this._client.onerror = (e) => {
                console.log('Connection Error', e)
                vscode.window.showErrorMessage('Failed to connect MM server. MM will retry after 3 sec.')
                if (this._retryLoop) {
                    clearTimeout(this._retryLoop)
                    this._retryLoop = null
                }
                this._retryLoop = setTimeout(() => {
                    this.startConnect()
                }, 1000 * 3)
            }
            this._client.onopen = () => {
                console.log('WebSocket Client Connected')
                vscode.window.showInformationMessage('MM Client Connected')
                this._client.send('join ' + JSON.stringify({
                    name: 'room'
                }))
                this._putLatestFile()
            }
            this._client.onclose = () => {
                console.log('WebSocket Client Closed')
            }
            this._client.onmessage = (e) => {
                const data = JSON.parse(e.data)
                const type = data.type
                if (type === 'join') {
                    // 誰か入室
                    this._putLatestFile()
                }
            }

            this._setStateStop()
        }

        private _setStateRun() {
            this._statusBarItem.text = "Run MM"
            this._statusBarItem.command = "extension.mmCodeStart"
        }
        
        private _setStateStop() {
            this._statusBarItem.text = "Stop MM"
            this._statusBarItem.command = "extension.mmCodeStop"
        }

        private _getFileData() {
            // アクティブなエディタを取得
            const editor = window.activeTextEditor
            return {
                fileName: editor ? editor.document.fileName : '',
                text: editor ? editor.document.getText() : ''
            }
        }

        private _putLatestFile(id = "") {
            const data = this._getFileData()
            this._client.send('follow ' + JSON.stringify({
                id: id,
                fileName: data.fileName,
                text: data.text,
            }))
            this._putLine()
        }
        
        private _putLine() {
            // アクティブなエディタを取得
            const editor = window.activeTextEditor
            if (editor) {
                const lineIndex = editor.selection.anchor.line
                const columnIndex = editor.selection.anchor.character
                const lineText = editor.document.lineAt(lineIndex).text
    
                this._client.send('line ' + JSON.stringify({
                    r: lineIndex,
                    c: columnIndex,
                    l: lineText
                }))
            }
        }

        private _onDidChangeTextDocument(e1: vscode.TextDocumentChangeEvent) {
            const updates = []
            e1.contentChanges.forEach((e2: vscode.TextDocumentContentChangeEvent) => {
                updates.push({
                    sr: e2.range.start.line,
                    sc: e2.range.start.character,
                    er: e2.range.end.line,
                    ec: e2.range.end.character,
                    t: e2.text
                })
            })

            this._client.send('updates ' + JSON.stringify({
                updates: updates
            }))
        }
    
        private _onDidChangeActiveTextEditor(e: vscode.TextEditor) {
            this._putLatestFile()
        }
        
        private _onDidChangeTextEditorSelection(e: vscode.TextEditorSelectionChangeEvent) {
            this._putLine()
        }
    }